src/uct/ib/base/ib_iface.c

```diff
@@ -506,14 +522,16 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
     }
 
     /* Address scope and size */
-#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
-    if (uct_ib_iface_port_attr(self)->link_layer == IBV_LINK_LAYER_ETHERNET) {
-        self->addr_type = UCT_IB_ADDRESS_TYPE_ETH;
-       } else
-#endif
-       {
-        self->addr_type = uct_ib_address_scope(self->gid.global.subnet_prefix);
-       }
+    if (config->addr_type == UCT_IB_IFACE_ADDRESS_TYPE_AUTO) {
+        if (IBV_PORT_IS_LINK_LAYER_ETHERNET(uct_ib_iface_port_attr(self))) {
+            self->addr_type = UCT_IB_ADDRESS_TYPE_ETH;
+        } else {
+            self->addr_type = uct_ib_address_scope(self->gid.global.subnet_prefix);
+        }
+    } else {
+        ucs_assert(config->addr_type < UCT_IB_ADDRESS_TYPE_LAST);
+        self->addr_type = config->addr_type;
```

Comment:
shouldn't we verify the user's choice? if, for example, the user chose "eth" on an IB port, it will fail since incorrect  packing and unpacking would be used.